File: extensions/oh-my-gemini/scripts/pre_tool.py
#!/usr/bin/env python3
import json
import os
import sys
from pathlib import Path

WRITE_TOOLS = {"write_file", "edit_file"}


def get_current_task(workspace_path: str):
    if not workspace_path:
        return None

    current_task_file = (
        Path(workspace_path) / ".agent_working_dir" / "current_task.json"
    )

    if not current_task_file.exists():
        return None

    try:
        with open(current_task_file, "r") as f:
            data = json.load(f)
        return data.get("current")
    except:
        return None


def is_sisyphus_path(file_path: str) -> bool:
    sisyphus_dirs = {".sisyphus", ".agent_working_dir", ".gemini"}
    path = Path(file_path)
    for part in path.parts:
        if part in sisyphus_dirs:
            return True
    return False


def main():
    workspace_path = os.environ.get("WORKSPACE_PATH", "")

    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        print(json.dumps({"decision": "allow"}), flush=True)
        return

    tool_name = input_data.get("tool_name", "")

    if tool_name not in WRITE_TOOLS:
        print(json.dumps({"decision": "allow"}), flush=True)
        return

    current_task = get_current_task(workspace_path)

    if not current_task:
        print(json.dumps({"decision": "allow"}), flush=True)
        return

    tool_input = input_data.get("tool_input", {})
    file_path = tool_input.get("file_path", "") or tool_input.get("file", "")

    if file_path and not is_sisyphus_path(file_path):
        additional_context = """

---

**[oh-my-gemini] ORCHESTRATOR REMINDER**

You are about to directly modify a file outside planning directories.

As an ORCHESTRATOR:
- Consider DELEGATING to a subagent instead
- If you must modify directly, keep it minimal
- After modifying, update **progress.md** with what was done

**Delegation is preferred for substantial work.**
"""

        print(
            json.dumps(
                {
                    "decision": "allow",
                    "hookSpecificOutput": {"additionalContext": additional_context},
                }
            ),
            flush=True,
        )
    else:
        print(json.dumps({"decision": "allow"}), flush=True)


if __name__ == "__main__":
    main()

File: extensions/oh-my-gemini/scripts/pre_tool.js
#!/usr/bin/env node
/**
 * @license
 * Copyright 2026 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * PreToolUse Hook
 * 
 * Reminds orchestrator to delegate or update progress when modifying files.
 */

const fs = require('fs');
const path = require('path');

const WORKSPACE_PATH = process.env.WORKSPACE_PATH || '';
const currentTaskFile = path.join(WORKSPACE_PATH, '.agent_working_dir', 'current_task.json');

const WRITE_TOOLS = ['write_file', 'edit_file'];

/**
 * Get current task directory
 * @returns {string|null} - Task directory path or null
 */
function getCurrentTask() {
  if (!WORKSPACE_PATH) {
    return null;
  }

  if (!fs.existsSync(currentTaskFile)) {
    return null;
  }

  try {
    const data = JSON.parse(fs.readFileSync(currentTaskFile, 'utf-8'));
    return data.current || null;
  } catch (err) {
    return null;
  }
}

/**
 * Check if path is in sisyphus directory
 * @param {string} filePath - File path to check
 * @returns {boolean} - True if in sisyphus directory
 */
function isSisyphusPath(filePath) {
  const sisyphusDirs = ['.sisyphus', '.agent_working_dir', '.gemini'];
  const pathParts = filePath.split(path.sep);
  for (const part of pathParts) {
    if (sisyphusDirs.includes(part)) {
      return true;
    }
  }
  return false;
}

/**
 * Output JSON to stdout
 * @param {object} data - Data to output
 */
function outputJson(data) {
  console.log(JSON.stringify(data));
}

function main() {
  // Read stdin
  let input_data = {};
  try {
    const stdin = fs.readFileSync(0, 'utf-8');
    input_data = JSON.parse(stdin);
  } catch (err) {
    outputJson({ decision: 'allow' });
    return;
  }

  const tool_name = input_data.tool_name || '';

  if (!WRITE_TOOLS.includes(tool_name)) {
    outputJson({ decision: 'allow' });
    return;
  }

  const currentTask = getCurrentTask();

  if (!currentTask) {
    outputJson({ decision: 'allow' });
    return;
  }

  const tool_input = input_data.tool_input || {};
  const file_path = tool_input.file_path || tool_input.file || '';

  if (file_path && !isSisyphusPath(file_path)) {
    const additional_context = `

---

**[oh-my-gemini] ORCHESTRATOR REMINDER**

You are about to directly modify a file outside planning directories.

As an ORCHESTRATOR:
- Consider DELEGATING to a subagent instead
- If you must modify directly, keep it minimal
- After modifying, update **progress.md** with what was done

**Delegation is preferred for substantial work.**
`;

    outputJson({
      decision: 'allow',
      hookSpecificOutput: { additionalContext: additional_context },
    });
  } else {
    outputJson({ decision: 'allow' });
  }
}

main();
